SimpleWAWS/Authentication/TwitterAuthProvider.cs

```diff
@@ -19,7 +19,12 @@ public override string GetLoginUrl(HttpContextBase context)
             builder.AppendFormat("&redirect_uri={0}", WebUtility.UrlEncode(string.Format(CultureInfo.InvariantCulture, "https://{0}/", context.Request.Headers["HOST"])));
             builder.AppendFormat("&client_id={0}", "");
             builder.AppendFormat("&scope={0}", "email");
-            builder.AppendFormat("&state={0}", WebUtility.UrlEncode(context.IsAjaxRequest() ? string.Format(CultureInfo.InvariantCulture, "/{0}{1}", culture, context.Request.Url.Query) : context.Request.Url.PathAndQuery));
+            if (context.IsFunctionsPortalRequest())
+            {
+                builder.AppendFormat("&state={0}", WebUtility.UrlEncode(string.Format(CultureInfo.InvariantCulture, "{0}{1}", context.Request.Headers["Referer"], context.Request.Url.Query)));
+            }
+            else
+                builder.AppendFormat("&state={0}", WebUtility.UrlEncode(context.IsAjaxRequest() ? string.Format(CultureInfo.InvariantCulture, "{0}{1}", culture, context.Request.Url.Query) : context.Request.Url.PathAndQuery));
```

Comment:
also there seem to be quite a bit of code duplication in here. If possible try moving common code into a function or the base class.